dasharo-security/sort-pci-semiauto.robot

```diff
@@ -0,0 +1,79 @@
+*** Settings ***
+Library             Collections
+Library             DateTime
+Library             OperatingSystem
+Library             Process
+Library             String
+Library             Telnet    timeout=20 seconds
+Library             SSHLibrary    timeout=90 seconds
+Library             RequestsLibrary
+Resource            ../rtectrl-rest-api/rtectrl.robot
+Resource            ../variables.robot
+Resource            ../keywords.robot
+
+Suite Setup         Run Keywords
+...                     Prepare Test Suite
+Suite Teardown      Run Keyword
+...                     Log Out And Close Connection
+
+
+*** Variables ***
+${LSPCI_CMD}=       lspci -vv
+${IP_LINK_CMD}=     ip link
+
+
+*** Test Cases ***
+Sort000.000 Port Order and PCIe Switching
+    [Documentation]    This test automates the verification of port order based on PCIe bus numbers and checks PCIe switching.
+
+    Power On
+    Boot System Or From Connected Disk    ubuntu
+    Login To Linux
+    Switch To Root User
+    ${pci_devices}=    Get PCIe Bus Info With MACs
+    Log    PCIe devices and their MACs: ${pci_devices}
+    Compare Interface And PCIe Bus Order    ${pci_devices}
+
+
+*** Keywords ***
+Get PCIe Bus Info With MACs
+    [Documentation]    Extract PCIe bus numbers and MAC addresses from lspci output.
+    ${lspci_output}=    Execute Command In Terminal    ${LSPCI_CMD}
+    ${lines}=    Split String    ${lspci_output}    \n
+    ${pci_devices}=    Create Dictionary
+    ${current_device}=    Set Variable    None
+    FOR    ${line}    IN    @{lines}
+        ${is_device_line}=    Evaluate    'Ethernet controller' in '${line}'
```

Comment:
There might be multiple Ethernet controllers in the system of different type. It is the case for VP66xx at least. To be specific, we may check for a specific controller, which could be defined by some variable. If the variable is not found, then we may look for `Ethernet controller` as a fallback.

Reply:
I think you don't understand... Let me illustrate that for you. This is what I do for an lspci output from VP66:

```
cat lspci.log |grep 'Ethernet controller'
02:00.0 Ethernet controller [0200]: Intel Corporation Ethernet Controller X710 for 10GbE SFP+ [8086:1572] (rev 02)
02:00.1 Ethernet controller [0200]: Intel Corporation Ethernet Controller X710 for 10GbE SFP+ [8086:1572] (rev 02)
03:00.0 Ethernet controller [0200]: Intel Corporation Device [8086:125c] (rev 04)
04:00.0 Ethernet controller [0200]: Intel Corporation Device [8086:125c] (rev 04)
05:00.0 Ethernet controller [0200]: Intel Corporation Device [8086:125c] (rev 04)
06:00.0 Ethernet controller [0200]: Intel Corporation Device [8086:125c] (rev 04)
```

As you can see there are two SFP ports, which are labeled separately from the ethernet port on chassis. We only want to check the ethernet ones, so you have to actually check for specific devices.

As a last resort you may add the SFP MACs as well to the config,  but then can we say that the test does what it is supposed to do? No, because it checks ports which are not supposed to be checked. There may be machines with different ethernet chips, which may not be internal (externally connected cards).

So what the test needs to do is t additionally check for the device name or ID which is of interest to us. In the example above it would be `8086:125c`

Reply:
I just added the SFP because they and regular ethernet looked alike, no wonder i got confused. Thanks for your extra time to explain this. I guess I'll just skip the SFP ones and this should be it: https://github.com/Dasharo/open-source-firmware-validation/pull/515/commits/8cdab97d10bb5f5312cdfc8926f3d8607ca704ae

Reply:
> I just added the SFP because they and regular ethernet looked alike, no wonder i got confused.

Yes, there are SFP to Ethernet adapters plugged in there, thus it looks the same... We do not have many means to test fibre connection, thus as a workaround we use the Ethernet plugs.

Reply:
> I guess I'll just skip the SFP ones and this should be it

Yes and no... Liek I have mentioned, you may find different network cards in the system, e.g. some external Ethernet Controllers which may not be SFP. So the test may not always catch this. To be sure, I would like the test to check for the device ID we are interested in, like in the comment above.

Reply:
ok i see, all protectli v1000 and VP6670 has the same ethernet controller id (8086:125c), this test looks for it and ignores any other line: https://github.com/Dasharo/open-source-firmware-validation/pull/515/commits/aef36b8ddc07b10943d7e4b55cc4f1d9c7f11984

Reply:
Yes, we can drop it